packages/circuit-ui/components/ModalContext/ModalContext.tsx

```diff
@@ -0,0 +1,206 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, {
+  createContext,
+  useContext,
+  useEffect,
+  useMemo,
+  useCallback,
+  useDebugValue,
+  FC,
+  MouseEvent,
+  KeyboardEvent,
+} from 'react';
+import ReactModal, { Props as ReactModalProps } from 'react-modal';
+import { Global, css } from '@emotion/core';
+// import noScroll from 'no-scroll';
+import { Dispatch as TrackingProps } from '@sumup/collector';
+
+import { useStack, StackItem, StackDispatch } from '../../hooks/useStack';
+import { uniqueId } from '../../util/id';
+// import IS_IOS from '../../util/ios';
+
+// TODO: Add explainer what this does.
+if (typeof window !== 'undefined') {
+  // These are the default app elements in Next.js and CRA.
+  const appElement =
+    document.getElementById('__next') || document.getElementById('root');
+
+  if (appElement) {
+    ReactModal.setAppElement(appElement);
+  } else if (process.env.NODE_ENV !== 'production') {
+    // TODO: Add error message
+    console.error('');
+  }
+}
+
+export type OnClose = (event?: MouseEvent | KeyboardEvent) => void;
+
+export interface BaseModalProps
+  extends Omit<
+    ReactModalProps,
+    'shouldCloseOnOverlayClick' | 'shouldCloseOnEsc'
+  > {
+  /**
+   * Callback function that is called when the modal is closed.
+   */
+  onClose?: OnClose;
+  /**
+   * Additional data that is dispatched with the tracking event.
+   */
+  tracking?: TrackingProps;
+}
+
+export type ModalComponent<T extends BaseModalProps> = ((
+  props: T,
+) => JSX.Element) & { TIMEOUT?: number };
+
+type ModalState = Omit<BaseModalProps, 'isOpen'> &
+  StackItem & { component: ModalComponent<BaseModalProps> };
+
+type ModalContextValue = [ModalState[], StackDispatch<ModalState>];
+
+const ModalContext = createContext<ModalContextValue>([[], () => {}]);
+
+export interface ModalProviderProps extends Omit<ReactModalProps, 'isOpen'> {
+  initialState?: ModalState[];
+}
+
+export const ModalProvider: FC<ModalProviderProps> = ({
+  children,
+  initialState,
+  portalClassName = 'ReactModalPortal',
+  htmlOpenClassName = 'ReactModal__Html--open',
+  ...defaultModalProps
+}) => {
+  const [modals, dispatch] = useStack<ModalState>(initialState);
+
+  const isOpen = modals.length > 0;
+
+  useEffect(() => {
+    const popModal = () => {
+      dispatch({ type: 'pop' });
+    };
+
+    if (isOpen) {
+      window.onpopstate = popModal;
+    } else {
+      window.onpopstate = null;
+    }
+
+    return () => {
+      window.onpopstate = null;
+    };
+  }, [dispatch, isOpen]);
+
+  // TODO: Not sure this even works or is still needed.
```

Comment:
I guess not, since you're handling the logic to prevent scroll differently

Reply:
Different browsers need different approaches. This was intended for Safari although the README for the dependency states that it doesn't work in Safari 🤷🏻‍♂️